### DISCUSSION TITLE
[Feature] Convert or share preview images when sharing

### DISCUSSION THREAD

**Discussion by NicholasFlamy** [immich-app/immich#6723] [Feature Request]
### The Feature

This issue is primarily focused on behavior when hitting "Share To" in the mobile app 
![Screenshot_20240129_084921](https://github.com/immich-app/immich/assets/30300649/61f8935c-3c76-42c5-a2c4-1d819b97f990)

When clicking the share button described above, an option to share the preview image instead of the original could be presented.

For web a Download Preview button could be added.
Note for web: you can right click an image and click save to get the preview image. (If you have NOT zoomed in and if you have view originals disabled (disabled by default).)

### Considerations

- [ ] Popup that when you select preview or original brings you to the Android share menu
    - Currently "Share To" goes straight to the Android share menu
- [ ] Option in app settings
    - Default to off because this would be an extra step (catered to more advanced users) in the share process
- [ ] Work offline if the preview image was cached
- [ ] Web preview download/sharing

<details><summary>Old Text</summary>
So two possible server-side implementations are that it tells the server to convert on demand or it converts and stores a second copy for every image uploaded (so it will create a second copy of a heic (or any photo) as a like 80% quality JPEG (potentially also WEBP), should be very configurable to the user/admin.)
Personally I think this solution is potentially inefficient, potentially slow (if you have a local file it has to be remotely transcoded and sent over network), and potentially complicated to implement. Maybe the on demand server-side conversion would be good, because of on device conversion potential issues. 

Then the option I might like is in-app (and maybe the web interface) conversion on sharing, so it can convert to whatever specified when you hit share. This could be complex since it would have to work for iOS and Android, and potentially the Web as well so there's a suite of potential difficulties supporting this for different devices.

Also, either route should allow deep configuration with potentially just a FFmpeg command, and/or an easy interface (kinda like the video transcoding tab or the thumbnails tab on settings.)
Personally, I would also like a profile selector, so when I hit share, I can choose a profile to convert to (so maybe have a default profile "Original" which does nothing and I could make a custom profile and name it whatever I want like "JPEG 80%" and make it convert to 80% quality JPEG) and then it pops up the share menu. Also, it should allow setting a profile to "Always Use" (of course on a per user basis) so that for example I (advanced user) could choose what to convert to, but an elderly family member on iPhone could have it automatically convert to JPEG without her having to even think about it.

So to sum up, the requirements I would like this feature to meet are:
1. Convert when you hit "Share To"
2. Configure the quality

Lower priority (but still much wanted):

1. Profiles for converting (making profiles with quality settings, also an original photo profile that doesn't convert) (created by admin and everyone can see the profiles if server side)
2. If it has profiles, setting a profile as "Always Use" (per User) so that you don't have to choose every time if you know you always want it to convert.
3. Multiple formats to convert to, maybe WEBP as an option.
4. Potentially an (under a toggle or drop-down) advanced config that allows something like an FFmpeg command

(Of course I'm not demanding y'all do this, I'm aware this is only a request, unlike some of the rude people who come here to demand something in their entitlement, just wanted to lay out some information that I would find helpful and maybe others can give their ideas.)

Also, if someone could point me to how I could help add this, I would be glad to try. I might be able to do some basic UI at least. (But I'm not skilled enough yet to implement this.)
</details>

### Platform

- [ ] Server
- [x] Web
- [X] Mobile

**Comment by NicholasFlamy** [comment]
Video is much less important because I assume most people use a share link. It might still be nice but it's harder to do on-demand and cut the file size without destroying the quality. I think most would use a link, but for images, sharing a single image is much more convenient with the file than a link. For example, on Discord, HEIC doesn't show a preview or anything so you have to download it, annoying and many people won't want to bother unless it's really important, share links might embed but usually not great as a way to view the picture, JPEG files show up and you can see it previewed and tap on it to view it, it's the intended method.
A video isn't something you glance over unless it's only like <=15 seconds but those are usually within Discord (and other platform) limitations whereas a long video requires some time to watch so clicking a link is not such a bother.

**Comment by mertalev** [comment]
How do you feel about using the preview image for this instead of converting on-the-fly? I think it'd be simpler to implement this way.

**Reply by NicholasFlamy** [reply]
I was just thinking about this. Maybe when you hit share to there is an option for either the image or the thumbnail or this option could be in the app settings.

**Reply by NicholasFlamy** [reply]
But, I think this should not be the permanent solution. I might want to share the higher quality version but simply as JPEG rather than HEIC without having my thumbnail quality having to be too high that it just wastes space and resources.

**Reply by mertalev** [reply]
I think using the preview would be the 80/20 solution here. The main hurdle for doing it locally is that we'd have to set up image processing code and libraries on mobile. We have a very specific setup on the server for this, so it's best to reuse that where possible.

Setting a custom quality when sharing is a valid, but separate request that would ideally still use the server to generate it.

**Reply by NicholasFlamy** [reply]
I think the on-device conversion can be mostly scrapped. The only reason it might come about is if we get editing features like cropping, rotating, drawing, etc. At least with drawing it seems like client-side would be easier than trying to pass all the changes to the server and process remotely. But yeah, server conversion is probably best.

**Reply by NicholasFlamy** [reply]
Putting this here so we know at least one other user wants this feature:

> My use case is that sometimes I don't have a connection to my Immich server but I have downloaded the photo previews on my "offline" instance of the Immich on my Android phone. I would like to still be able to share the photo version that I have (preview in this case) to other apps on my phone, e.g. for editing or just send it to someone (it will be downscaled by messaging apps anyway so preview version is perfect for this). I don't see why it should be necessary to download the original one from the server for this, especially if in settings I've selected to only download the photo previews to my device.

_Originally posted by @kamzil in https://github.com/immich-app/immich/discussions/1863#discussioncomment-8456580_

**Comment by NicholasFlamy** [comment]
> server to convert on demand

After thinking about it for some time, this seems like the best solution. No worry about supporting conversion on every platform and works well with images on the server (so basically all images since that's the point of immich) and should work fine on the web (which doesn't have local images, all on server.)

**Comment by NicholasFlamy** [comment]
Anyone know how tf this custom sharing menu in YouTube works? I was wondering about the possibility of immich having this (optionally) so that future features such as my proposed idea to convert on sharing could have per app profile settings.
![Screenshot_20240129_194743_ReVanced Extended](https://github.com/immich-app/immich/assets/30300649/5ec211d5-5029-462b-93cf-abaa26b8443c)

**Comment by BoBBer446** [comment]
That would be a great thing.

I go out a lot and take photos with my DSL. These are in .ARW format. So I import them into Immich on the computer.

But now when I look at the photos in the Android app and want to share one, it doesn't work if it's an .ARW. 

At least it would be advantageous to have a choice here.

**Comment by NicholasFlamy** [comment]
For now I have a partial solution and that is ImagePipe. Basically you set it up and then share image(s) to ImagePipe and it converts and then pops up a share menu so you can share the converted image(s) to another app

**Reply by BoBBer446** [reply]
Works Not well but it is a way, true. Thank you for Sharing

**Comment by kamzil** [comment]
This would allow more similar functionality to sharing images via IM apps like WhatsApp and Signal, where by default images are shared as low-quality (in Immich case, default should still probably be orig quality as it's now). It's a great way to save on bandwidth, and would allow sharing photos even when you don't have connection to the server, as long as they've been cached on your device

**Comment by NicholasFlamy** [comment]
With the release of https://github.com/immich-app/immich/releases/tag/v1.131.0 we now have the option for creating a full resolution JPEG version of images. Maybe this could be considered for this feature.

**Reply by NicholasFlamy** [reply]
@mertalev I want to hear your thoughts on using this new feature (and if disabled using the preview). More specifically, I would like to hear how you (or other team members) feel is the best way to implement this option.

Two possibilities I can think of are prompting the user asking which one to share after they click share, and the other is a mobile app or web app setting to always pick the "Full-size Image" (and if disabled, the preview).

**Reply by AngelaDMerkel** [reply]
I am not a team member, but I would like to weigh in on the side of providing an admin controlled default and offer sharing the JPEG thumb or if https://github.com/immich-app/immich/discussions/7666 is ever implemented use that to generate the image according to a default. 

I store my files in JXL to reduce storage demands, but they are useless for sharing, unfortunately.

**Reply by WojtaCZ** [reply]
I'd also very much appreciate this feature. I have a large library of high resolution TIFF film scans. These take ages to be share using the share button (sometimes gigabytes to be downloaded) and it!s nonsence to share the high res originals. In my opinion, the functionality to automatically transcode images to lower resolution and quality for sharing is a must have...
